services/src/pro/datasets/postgres.rs

```diff
@@ -745,36 +794,46 @@ where
         let mut conn = self.conn_pool.get().await?;
         let tx = conn.build_transaction().start().await?;
 
-        self.ensure_permission_in_tx(dataset_id.into(), Permission::Owner, &tx)
-            .await
-            .boxed_context(crate::error::PermissionDb)?;
-
-        let stmt = tx
-            .prepare(
-                "
-        SELECT 
-            TRUE
-        FROM 
-            user_permitted_datasets p JOIN datasets d 
-                ON (p.dataset_id = d.id)
-        WHERE 
-            d.id = $1 AND p.user_id = $2 AND p.permission = 'Owner';",
-            )
-            .await?;
-
-        let rows = tx
-            .query(&stmt, &[&dataset_id, &self.session.user.id])
-            .await?;
-
-        if rows.is_empty() {
-            return Err(Error::OperationRequiresOwnerPermission);
+        let _uploaded = self.uploaded_dataset_status_in_tx(&dataset_id, &tx).await;
+        if let Err(error) = _uploaded {
+            if matches!(error, UnknownDatasetId) {
+                self.ensure_permission_in_tx(dataset_id.into(), Permission::Owner, &tx)
+                    .await
+                    .boxed_context(crate::error::PermissionDb)?;
+
+                let stmt = tx
+                    .prepare(
+                        "
+                SELECT
+                    TRUE
+                FROM
+                    user_permitted_datasets p JOIN datasets d
+                        ON (p.dataset_id = d.id)
+                WHERE
+                    d.id = $1 AND p.user_id = $2 AND p.permission = 'Owner';",
+                    )
+                    .await?;
+
+                let rows = tx
+                    .query(&stmt, &[&dataset_id, &self.session.user.id])
+                    .await?;
+
+                if rows.is_empty() {
+                    return Err(Error::OperationRequiresOwnerPermission);
+                }
+
+                let stmt = tx.prepare("DELETE FROM datasets WHERE id = $1;").await?;
+
+                tx.execute(&stmt, &[&dataset_id]).await?;
+
+                tx.commit().await?;
+
+                return Ok(());
+            }
         }
 
-        let stmt = tx.prepare("DELETE FROM datasets WHERE id = $1;").await?;
-
-        tx.execute(&stmt, &[&dataset_id]).await?;
-
-        tx.commit().await?;
+        self.expire_uploaded_dataset(ChangeDatasetExpiration::delete_full(dataset_id))
```

Comment:
is every delete an expiration now?

Reply:
If it is a naming thing, I'm...open either way.

From a technical standpoint, the current implementation interprets _delete_ as _expire now_, which means there is only one internal logic (i.e., set of queries) for this process. I could also change that.